Compaction lossy hash

dotnwat opened the pull request:
Introduces a new offset key map implementation that maps compaction key space into sha256(key) space.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v23.2.x
- [ ] v23.1.x
- [ ] v22.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->
* none

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/39629#018b5ebb-73a0-46d9-82bd-b910bbd31174

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/39629#018b5ebb-73a4-4727-af38-9ae04aeb92d8

emaxerrno commented:
i was skeptical at first w/ the regular byte comparison hash. this makes sense.

VladLazar commented on src/v/utils/fragmented_vector.h:
```diff
@@ -222,6 +222,57 @@ class fragmented_vector {
+            if (_size == _capacity) {
+                std::vector<T> frag;
+                if (deficit >= elems_per_frag) {
+                    frag.resize(elems_per_frag);
```
nit: not sure if `std::vector::resize` has a guarantee that `capacity == size` after a call which requires increasing capacity

dotnwat replied:
Could you expand on the concern you have? Is the concern that `capacity != size` related a correctness concern, or an optimization concern (i guess or both)?

VladLazar replied:
My concern is that if `resize` tries to be smart and increases the capacity more than strictly required, then `_capacity` would become out of sync with the actual capacity. Played around with this and it doesn't seem to be the case though.

dotnwat replied:
>capacity more than strictly required, then _capacity would become out of sync

got it. i don't think we ever inspect the capacity of the underlying vectors, so it sounds like your concern is optimization rather than out-of-sync affecting correctness?

>My concern is that if resize tries to be smart and increases the capacity more than strictly required

Was just reading the implementation in libc++. It does seem to double the capacity when resize() appends new entries. However, since that adaptive capacity feature operates on capacity and capacity of a new vector is zero, then resize() on a empty vector does the right thing and uses a precise allocation size.

Interesting!

VladLazar replied:
Yeah, I don't think it impacts correctness, but would be surprising. I think you can always do a `shrink_to_fit` call after the resize, but maybe it doesn't matter that much.

dotnwat replied:
I just added a reserve() and a note. In this case, at least in libc++ resize won't do any new allocations if the capacity is sufficient.

dotnwat replied:
I ended up removing frag_vec::resize i think it was a bit awkward

VladLazar commented on src/v/utils/fragmented_vector.h:
```diff
@@ -222,6 +222,57 @@ class fragmented_vector {
+            }
+
+            const auto add_count = std::min(deficit, _capacity - _size);
+            _frags.back().resize(_frags.back().size() + add_count);
```
nit: could resize to `elems_per_frag` and skip the arithmetic

dotnwat replied:
doesn't that only work if `deficit` is larger than the remaining space in the fragment? what if the fragment was half full but `deficit` was `1`?

VladLazar replied:
Yeah, you're right. Makes sense

VladLazar commented on src/v/storage/key_offset_map.h:
```diff
@@ -91,4 +93,81 @@ class simple_key_offset_map final : public key_offset_map {
+        bool empty() const;
+    };
+
+    /**
+     * Uses successive chunks of sizeof(index_type) bytes taken from hash(key)
+     * as probes into the hash table. When `next()` returns null then the caller
+     * should switch to linear probing.
+     */
+    struct probe {
+        using index_type = uint32_t;
+        static_assert(sizeof(index_type) <= hash_type::digest_size);
+
+        explicit probe(const hash_type::digest_type&);
+
+        std::optional<index_type> next();
+
+        hash_type::digest_type::const_pointer iter;
+        hash_type::digest_type::const_pointer end;
+    };
```
So the idea here is to probe at random places in the backing vector. This is neat. Does this smartness have a name?

dotnwat replied:
I looked extensively to see if this had a name, but I don't think it does.

IMO this is probably just a different flavor/generalization of _double hashing_ where if the first hash position is a collision a different hash function is used as the next probe. Here the hash output is large enough that we wouldn't use all those bits anyway for the first position.

VladLazar commented on src/v/storage/key_offset_map.cc:
```diff
@@ -49,4 +49,182 @@ model::offset simple_key_offset_map::max_offset() const { return _max_offset; }
+            ++size_;
+            max_offset_ = std::max(max_offset_, offset);
+            return true;
+        } else if (entry.digest == hash) {
```
Have we ruled out collisions as something we don't need to worry about?

VladLazar commented on src/v/storage/key_offset_map.cc:
```diff
@@ -49,4 +49,182 @@ model::offset simple_key_offset_map::max_offset() const { return _max_offset; }
+    while (!entries_.empty()) {
+        entries_.pop_back_n(
+          std::min(entries_.size(), entries_.elements_per_fragment()));
+        if (seastar::need_preempt()) {
```
this is cool.

dotnwat replied:
Yeh. Seems to address the performance cost of the co_await in a tight loop.

VladLazar commented on src/v/storage/key_offset_map.cc:
```diff
@@ -49,4 +49,182 @@ model::offset simple_key_offset_map::max_offset() const { return _max_offset; }
+    }
+
+    entries_.clear();
+    while (entries_.memory_size() < size) {
+        entries_.resize(entries_.size() + entries_.elements_per_fragment());
+        if (seastar::need_preempt()) {
+            co_await seastar::maybe_yield();
+        }
+    }
```
Feels like this stuff should be encapsulated in the fragmented vector or in some utils namespace. Same for the "clear" above.

dotnwat replied:
Done. Added some helpers in the latest force push.

VladLazar reviewed (commented):
Very nice. A few questions/suggestions

andrwng reviewed (commented):
LGTM! Just some nits

dotnwat commented on tools/format-cc:
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
```
Nah, just a delivery vehicle for a small fix.

dotnwat commented on src/v/storage/tests/key_offset_map_test.cc:
```diff
@@ -0,0 +1,110 @@
+#include "storage/key_offset_map.h"
```
Fixed

dotnwat commented on src/v/utils/fragmented_vector.h:
```diff
@@ -222,6 +222,57 @@ class fragmented_vector {
+     * less than \p count then default-initialized elements are added. If the
+     * current size is larger then elements are removed.
+     *
+     * The expected use case for this is to allocate a large vector in a fiber
+     * using a series of smaller resize() invocations allowing for cooperative
+     * yield calls to be inserted to avoid reactor stalls. The optimal strategy
```
I don't think there are any magic numbers here. Since fragmented vector isn't futurized, it be the same concern as resizing a std::vector or avoiding calling fragmented_vector::copy on a large fragmented vector.

dotnwat commented on src/v/storage/key_offset_map.cc:
```diff
@@ -49,4 +49,182 @@ model::offset simple_key_offset_map::max_offset() const { return _max_offset; }
+    entries_.clear();
+    while (entries_.memory_size() < size) {
+        entries_.resize(entries_.size() + entries_.elements_per_fragment());
+        if (seastar::need_preempt()) {
```
It is, but there is a measurable, big performance cost calling co_await in a tight loop when the expectation is that the call should effectively be zero cost in most invocations.

dotnwat replied:
Fixed this by adding some helpers for interacting with the fragmented vector in the latest force push.

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/39695#018b6311-32e1-459d-91ec-ba8d198872d7

dotnwat commented:
force-push

* added cap on load factor at 95%
* enforced capacity on put()
* removed fragmented_vector::resize its a bit awkward
* added helper `fragmented_vector_clear_async`
* added helper `fragmented_vector_fill_async`
* added `hash_key_offset_map::initialize` for resetting hash table context that avoids reallocation with resize()
* switched to large fragmented vector variant

vbotbuildovich commented:
ducktape was retried in job https://buildkite.com/redpanda/redpanda/builds/39719#018b63ea-6c0d-4d9e-a600-2a2d69a8ae33

dotnwat commented:
Failure is https://github.com/redpanda-data/redpanda/issues/14218

andrwng commented on src/v/storage/key_offset_map.cc:
```diff
@@ -49,4 +49,209 @@ model::offset simple_key_offset_map::max_offset() const { return _max_offset; }
+    probe_count_ = 0;
+}
+
+seastar::future<> hash_key_offset_map::initialize() {
```
nit: maybe consider naming this `reset()` or something? As is it seems like something we should always call before using the map, but I don't think that's the case (presumably we just need to use `reset(size_bytes)`?

dotnwat replied:
`reset(size)` implies `initialize()`. but in practice you really only want to call `reset(size)` once at boot-up, and then call `initialize()` each time you want to use the table to do a scan because it avoids freeing and reallocating all the memory.

dotnwat commented on src/v/storage/key_offset_map.cc:
```diff
@@ -49,4 +49,209 @@ model::offset simple_key_offset_map::max_offset() const { return _max_offset; }
+    };
+
+    // handle a non-normalized probe position
+    // returns true if key is inserted
```
fixed
